pkg/virt-operator/util/readycheck.go

```diff
@@ -0,0 +1,169 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2019 Red Hat, Inc.
+ *
+ */
+
+package util
+
+import (
+	"strings"
+
+	appsv1 "k8s.io/api/apps/v1"
+	k8sv1 "k8s.io/api/core/v1"
+
+	v1 "kubevirt.io/kubevirt/pkg/api/v1"
+	"kubevirt.io/kubevirt/pkg/log"
+)
+
+func DaemonsetIsReady(kv *v1.KubeVirt, daemonset *appsv1.DaemonSet, stores Stores) bool {
+
+	// ensure we're looking at the latest daemonset from cache
+	obj, exists, _ := stores.DaemonSetCache.Get(daemonset)
+	if exists {
+		daemonset = obj.(*appsv1.DaemonSet)
+	} else {
+		// not in cache yet
+		return false
+	}
+
+	if daemonset.Status.DesiredNumberScheduled == 0 ||
+		daemonset.Status.DesiredNumberScheduled != daemonset.Status.NumberReady {
+
+		log.Log.V(4).Infof("DaemonSet %v not ready yet", daemonset.Name)
+		return false
+	}
+
+	// cross check that we have 'daemonset.Status.NumberReady' pods with
+	// the desired version tag. This ensures we wait for rolling update to complete
+	// before marking the infrastructure as 100% ready.
+	var podsReady int32
+	for _, obj := range stores.InfrastructurePodCache.List() {
+		if pod, ok := obj.(*k8sv1.Pod); ok {
+			if !podIsRunning(pod) {
+				continue
+			} else if !podHasNamePrefix(pod, daemonset.Name) {
+				continue
+			}
+
+			if !podIsUpToDate(pod, kv) {
+				log.Log.Infof("DaemonSet %v waiting for out of date pods to terminate.", daemonset.Name)
+				return false
+			}
+
+			if podIsReady(pod) {
+				podsReady++
+			}
+		}
+	}
+
+	if podsReady == 0 {
+		log.Log.Infof("DaemonSet %v not ready yet. Waiting on at least one ready pod", daemonset.Name)
+		return false
+	}
+
+	return true
+}
+
+func DeploymentIsReady(kv *v1.KubeVirt, deployment *appsv1.Deployment, stores Stores) bool {
+	// ensure we're looking at the latest deployment from cache
+	obj, exists, _ := stores.DeploymentCache.Get(deployment)
+	if exists {
+		deployment = obj.(*appsv1.Deployment)
+	} else {
+		// not in cache yet
+		return false
+	}
+
+	var specReplicas int32 = 1
+	if deployment.Spec.Replicas != nil {
+		specReplicas = *deployment.Spec.Replicas
+	}
+	if specReplicas != deployment.Status.Replicas ||
+		deployment.Status.Replicas != deployment.Status.ReadyReplicas {
+		log.Log.V(4).Infof("Deployment %v not ready yet", deployment.Name)
+		return false
+	}
+
+	// cross check that we have 'deployment.Status.ReadyReplicas' pods with
+	// the desired version tag. This ensures we wait for rolling update to complete
+	// before marking the infrastructure as 100% ready.
+	var podsReady int32
+	for _, obj := range stores.InfrastructurePodCache.List() {
```

Comment:
Here you can do the following things:
 1) Check if `status.observedGeneration` is equal to the `metadata.generation`
 2) `status.updatedReplicas` should be equal to `status.Replicas`
 3) `status.ReadyReplicas` should be equal to `status.Replicas`